nearcore/src/runtime/mod.rs

```diff
@@ -1542,6 +1580,73 @@ impl RuntimeAdapter for NightshadeRuntime {
         }
     }
 
+    fn build_state_for_split_shards(
+        &self,
+        shard_uid: ShardUId,
+        state_root: &StateRoot,
+        next_epoch_shard_layout: &ShardLayout,
+    ) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+        let trie = self.get_tries().get_view_trie_for_shard(shard_uid);
+        let shard_id = shard_uid.shard_id();
+        let mut state_roots = {
+            let split_shards = next_epoch_shard_layout
+                .get_split_shards(shard_id)
+                .ok_or(ErrorKind::InvalidShardId(shard_id))?;
+            split_shards
+                .iter()
+                .map(|id| {
+                    (
+                        ShardUId::from_shard_id_and_layout(*id, &next_epoch_shard_layout),
+                        StateRoot::default(),
+                    )
+                })
+                .collect()
+        };
+
+        let state_root_node = trie.retrieve_root_node(state_root).map_err(|err| err.to_string())?;
+        let num_parts = get_num_state_parts(state_root_node.memory_usage);
+        debug!("splitting state for shard {} to {} parts to build new states", shard_id, num_parts);
+        for part_id in 0..num_parts {
+            let trie_items = trie.get_trie_items_for_part(part_id, num_parts, state_root).map_err(|e|{
+                error!(target: "runtime",
+                       "Can't get_trie_nodes_for_part for {:?}, part_id {:?}, num_parts {:?}, {:?}",
+                       state_root, part_id, num_parts, e
+                );
+                e.to_string()
+            })?;
+            let (store_update, new_state_roots) = self
+                .tries
+                .apply_changes_to_new_states(
+                    state_roots,
+                    trie_items.into_iter().map(|(key, value)| (key, Some(value.clone()))).collect(),
+                    Box::new(|raw_key| {
+                        // Here changes on DelayedReceipts or DelayedReceiptsIndices will be excluded
+                        // This is because we cannot migrate delayed receipts part by part. They have to be
+                        // reconstructed in the new states after all DelayedReceipts are ready in the original
+                        // shard.
+                        if let Some(account_id) = parse_account_id_from_raw_key(&raw_key).unwrap() {
```

Comment:
is the unwrap safe?

Reply:
I changed it so that the error is propagated